src/_P044_P1WifiGateway.ino

```diff
@@ -5,127 +5,223 @@
 //
 //  designed for combo
 //    Wemos D1 mini (see http://wemos.cc) and
-//    P1 wifi gateway shield (see https://circuits.io/circuits/2460082)
-//    see http://romix.macuser.nl for kits
+//    P1 wifi gateway shield (see http://www.esp8266thingies.nl for print design and kits)
+//    See also http://domoticx.com/p1-poort-slimme-meter-hardware/
 //#######################################################################################################
 
 #include "_Plugin_Helper.h"
 
 #define PLUGIN_044
 #define PLUGIN_ID_044         44
 #define PLUGIN_NAME_044       "Communication - P1 Wifi Gateway"
-#define PLUGIN_VALUENAME1_044 "P1WifiGateway"
+
+#ifndef PLUGIN_044_DEBUG
+  #define PLUGIN_044_DEBUG                 false // extra logging in serial out
+#endif
 
 #define P044_STATUS_LED                    12
-#define P044_BUFFER_SIZE                   1024
+#define P044_CHECKSUM_LENGTH               4
+#define P044_DATAGRAM_START_CHAR           '/'
+#define P044_DATAGRAM_END_CHAR             '!'
+#define P044_DATAGRAM_MAX_SIZE             1024
 #define P044_NETBUF_SIZE                   128
-#define P044_DISABLED                      0
-#define P044_WAITING                       1
-#define P044_READING                       2
-#define P044_CHECKSUM                      3
-#define P044_DONE                          4
 
+#define P044_WIFI_SERVER_PORT     ExtraTaskSettings.TaskDevicePluginConfigLong[0]
+#define P044_BAUDRATE             ExtraTaskSettings.TaskDevicePluginConfigLong[1]
+#define P044_RX_WAIT              PCONFIG(0)
+#define P044_SERIAL_CONFIG        PCONFIG(1)
+#define P044_RESET_TARGET_PIN     CONFIG_PIN1
+ 
 
+struct P044_Task : public PluginTaskData_base {
 
-struct P044_data_struct : public PluginTaskData_base {
+  enum ParserState : byte {
+    P044_WAITING,
+    P044_READING,
+    P044_CHECKSUM
+  };
 
-  P044_data_struct() {
+  P044_Task() {
     clearBuffer();
   }
 
-  ~P044_data_struct() {
+  ~P044_Task() {
     stopServer();
   }
 
+  inline static bool serverActive(WiFiServer *server) {
+#if defined(ESP8266)
+    return nullptr != server && server->status() != CLOSED;
+#elif defined(ESP32)
+    return nullptr != server && *server;
+#endif
+  }
+
   void startServer(unsigned int portnumber) {
+    if (gatewayPort == portnumber && serverActive(P1GatewayServer)) {
+      // server is already listening on this port
+      return;
+    }
     stopServer();
+    gatewayPort = portnumber;
     P1GatewayServer = new WiFiServer(portnumber);
-    if (nullptr != P1GatewayServer) {
+    if (nullptr != P1GatewayServer && WiFi.isConnected()) {
```

Comment:
This function is only called during init, which is only called when you change settings, or boot.
So if the wifi is not yet connected when init of this task is called, the server will not be started.

You do check it periodically, so is it then still needed during init?
If you remove it from init you will run into the problem I will mention at the `checkServer` function :)

Reply:
Ok

Reply:
> I explicitly added the check for WiFi connected, not because the server would not start, but because the ESP32 tends to reboot when the WiFi becomes active.

Hmm that's an interesting observation.